### PR TITLE
601: gscan groups

### PR DESCRIPTION
**Pull request by arjclark**
Adds "group" config item to allow collapsing groups of suites in gscan.

A simple suite like this:

```
title = "Test Suite"
description="Testing grouping options"
group="test group"

[scheduling]
[[dependencies]]
graph = foo
```

will appear under a "Group" heading of "test group", if displayed in gscan.

@oliver-sanders - can you take an initial cursory look as someone else who's had to work with this recently? (also includes the indexing "fix" discussed)

Not ready for full review/merge yet.

### CODE REVIEW

**Comment by hjoliver** [comment]
I wonder if it would be better to define grouping in the user's `gscan.rc`, for the following reasons:
 * these groups only affect display in gscan (at least I can't think of other uses at this stage...)
 * you can see all group members at a glance (else, have to grep through all suite definition?)
 *easy to use this new functionality without editing existing (and potentially running) suites

**Comment by hjoliver** [comment]
This looks good, on a quick test.  I wonder if we could have a combined status summary for all group members though? (may be too difficult, just wondering...)

**Comment by arjclark** [comment]
> I wonder if we could have a combined status summary for all group members though

It's certainly something I'm planning on trying to do - there's a few bits that'll be a pain to do given how we extract status information from suites but I don't see it as not being possible.

**Comment by arjclark** [comment]
> I wonder if it would be better to define grouping in the user's gscan.rc, for the following reasons:
> * these groups only affect display in gscan (at least I can't think of other uses at this stage...)
> * you can see all group members at a glance (else, have to grep through all suite definitions?)
> * easy to use this new functionality without editing existing (and potentially running) suites

I can see both sides of this tbh. @cylc/core - any preferences either way and if we went down @hjoliver's route how you'd ideally like to see this configured?

**Comment by arjclark** [comment]
So, we've had a bit of a chat on our end (primarily me and @dpmatthews) and can see both arguments for where the configuration option should be. On the one hand, putting it in the suite.rc seems the natural place to put it and exposes the setting to the user rather than burying it away in a gcylc.rc file where it risks being forgotten about/not used and becomes an extra file to modify etc. On the other having it in the gcylc file would mean you could easily change the grouping for running suites after, say, spinning up a set of operational suites and then having to subsequently reload them to apply a change in grouping.

In the long term there's no reason why we couldn't support both with e.g. gcylc.rc providing default group values which then suite.rc entries could override. For now though, we'd prefer to go with the solution here and then, based on uptake/usage, we could extend to the gcylc.rc as well in a future pull request. Does this sound ok to you @hjoliver?

**Comment by hjoliver** [comment]
Yep, that's fine.  I like the idea of defaulting to settings in gscan.rc (if any), but that can be another PR.

**Comment by arjclark** [comment]
Coming back round to this now - been distracted elsewhere. Working on trying to get group summaries up and running.

**Comment by arjclark** [comment]
In theory all working now.

@oliver-sanders @hjoliver please review 1 and 2 respectively.

**Comment by oliver-sanders** [comment]
I'm getting this for the example in the original comment.

screenshot-untitled window

Am I doing something wrong?

**Comment by oliver-sanders** [comment]
`t/pep8` is failing.

**Comment by arjclark** [comment]
@oliver-sanders - sorry rebased and broke things.

**Comment by oliver-sanders** [comment]
I've got some strange issues with status icons appearing as black boxes also some (possibly un-related) traceback.

screenshot-cylc gscan-1

screenshot-untitled window-1

**Comment by benfitzpatrick** [comment]
Those black boxes are ‘unknown’ status, FYI

**Comment by arjclark** [comment]
@oliver-sanders - black box issue resolved, along with a few others.

**Comment by arjclark** [comment]
Cannot reproduce the traceback problem - I believe this isn't a result of this change though.

File: lib/cylc/gui/gscan.py [hunk 1]
```diff
@@ -1123,10 +1181,35 @@ def update(self):
             )
             title = suite_info.get("title")
 
+            group = suite_info.get("group")
+            if group is None:
+                group = ""
```
Comment (oliver-sanders) [inline lib/cylc/gui/gscan.py]: `group="Un-grouped"` might be more helpful.

File: lib/cylc/gui/gscan.py [hunk 2]
```diff
@@ -1144,27 +1227,31 @@ def update(self):
                         if state != TASK_STATUS_RUNAHEAD:
                             # 'runahead' states are usually hidden.
                             states_text += '%s %d ' % (state, number)
+                            # Extractable counts here?
```
Comment (oliver-sanders) [inline lib/cylc/gui/gscan.py]: Nope?

**Comment by oliver-sanders** [comment]
Moved traceback issue to #2061.

File: lib/cylc/gui/gscan.py (hunk 2 above)
Comment (arjclark) [inline lib/cylc/gui/gscan.py]: As it turned out, no

**Comment by arjclark** [comment]
Feedback from @oliver-sanders addressed. Over to you @hjoliver (or @matthewrmshin if you want to take it)

File: lib/cylc/gui/gscan.py [hunk 3]
```diff
@@ -1123,10 +1181,35 @@ def update(self):
             )
             title = suite_info.get("title")
 
+            group = suite_info.get("group")
+            if group is None:
+                group = "ungrouped"
```
Comment (matthewrmshin) [inline lib/cylc/gui/gscan.py]: I wonder if a string like `(default)` is better here?

File: lib/cylc/cfgspec/suite.py [hunk 4]
```diff
@@ -203,6 +203,7 @@ def _coerce_parameter_list(value, keys, _):
 SPEC = {
     'title': vdr(vtype='string', default=""),
     'description': vdr(vtype='string', default=""),
+    'group': vdr(vtype='string', default=""),
```
Comment (matthewrmshin) [inline lib/cylc/cfgspec/suite.py]: Worth adding a simple *validate* test for this new setting?

File: lib/cylc/gui/gscan.py (hunk 3 above)
Comment (oliver-sanders) [inline lib/cylc/gui/gscan.py]: I think "ungrouped" is clearer.

File: lib/cylc/cfgspec/suite.py (hunk 4 above)
Comment (arjclark) [inline lib/cylc/cfgspec/suite.py]: What did you have in mind?

**Comment by arjclark** [comment]
@matthewrmshin - default group is now "(ungrouped)". Validation test added.

**Comment by hjoliver** [comment]
(Nice work y'all)

**Pull request merged**